File: types/agent.ts
// Widget Configuration for embeddable chat
export interface WidgetConfig {
    title?: string;
    subtitle?: string;
    primaryColor?: string;
    position?: 'bottom-right' | 'bottom-left';
    logoUrl?: string;
    initialMessage?: string;
    showFooter?: boolean;
}


export interface SecuritySettings {
    enabled: boolean;
    fail_close?: boolean;
    check_jailbreak?: boolean;
    check_nsfw?: boolean;
    pii_action?: 'mask' | 'block' | 'off';
    check_secret_keys?: boolean;
    check_urls?: boolean;
    // URL Protection
    url_protection_mode?: 'off' | 'whitelist' | 'blacklist';
    url_whitelist?: string[];
    url_blacklist?: string[];
    allowed_topics?: string[];
    custom_regex?: string[];
    error_message?: string;
}

export interface Agent {
    id: string;
    company_id: string;
    name: string;
    slug: string;
    avatar_url?: string;
    is_active: boolean;

    // LLM Config
    llm_provider?: string;
    llm_model?: string;
    llm_temperature: number;
    llm_max_tokens: number;
    llm_top_p: number;
    llm_top_k: number;
    llm_frequency_penalty: number;
    llm_presence_penalty: number;

    // LLM Advanced Config (GPT-5.x, o1, o3)
    reasoning_effort?: string;
    verbosity?: string;

    // Behavior
    agent_system_prompt?: string;
    agent_enabled: boolean;
    use_langchain: boolean;

    // Capabilities
    allow_web_search: boolean;
    allow_vision: boolean;
    vision_model?: string;
    is_hyde_enabled?: boolean;

    // Tools Config (JSON)
    tools_config?: {
        human_handoff?: { enabled: boolean };
        [key: string]: { enabled: boolean } | undefined;
    };

    // Security Config (JSONB)
    security_settings?: SecuritySettings;

    // SubAgent Config
    is_subagent?: boolean;
    allow_direct_chat?: boolean;

    // Widget Config (JSONB)
    widget_config?: WidgetConfig;

    // Computed
    has_api_key: boolean;
    has_vision_api_key: boolean;
    has_whatsapp: boolean;

    // Timestamps
    created_at: string;
    updated_at: string;
}

File: backend/app/models/agent.py
from datetime import datetime
from typing import Any, Dict, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# --- BASE MODEL ---
class AgentBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    slug: str = Field(..., min_length=1, max_length=255)
    avatar_url: Optional[str] = None
    is_active: bool = True

    # LLM Config
    llm_provider: Optional[str] = None
    llm_model: Optional[str] = None
    llm_temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    llm_max_tokens: int = Field(default=2000, ge=100)
    llm_top_p: float = Field(default=1.0, ge=0.0, le=1.0)
    llm_top_k: int = Field(default=40, ge=1)
    llm_frequency_penalty: float = Field(default=0.0, ge=-2.0, le=2.0)
    llm_presence_penalty: float = Field(default=0.0, ge=-2.0, le=2.0)

    # Agent Behavior
    agent_system_prompt: Optional[str] = None
    agent_enabled: bool = True
    use_langchain: bool = False

    # Capabilities
    allow_web_search: bool = True
    allow_vision: bool = False
    vision_model: Optional[str] = None
    is_hyde_enabled: bool = True  # Deep RAG search with HyDE (slower but more accurate)

    # Strategic
    tools_config: Dict[str, Any] = Field(default_factory=dict)

    # Widget Config (JSONB - for embeddable chat widget)
    widget_config: Dict[str, Any] = Field(default_factory=dict)

    # LLM Advanced Config (GPT-5.x, o1, o3)
    reasoning_effort: Optional[str] = Field(default="medium")  # none, low, medium, high
    verbosity: Optional[str] = Field(default="medium")  # low, medium, high

    # Security Config (JSONB)
    security_settings: Dict[str, Any] = Field(default_factory=dict)

    # SubAgent Config
    is_subagent: bool = False              # Esconde widget/WhatsApp/canais no front
    allow_direct_chat: bool = False        # SubAgent aparece no chat test para treino


class AgentCreate(AgentBase):
    company_id: UUID  # Required to associate agent with a company

class AgentUpdate(BaseModel):
    name: Optional[str] = None
    slug: Optional[str] = None
    avatar_url: Optional[str] = None
    is_active: Optional[bool] = None

    llm_provider: Optional[str] = None
    llm_model: Optional[str] = None
    llm_temperature: Optional[float] = None
    llm_max_tokens: Optional[int] = None
    llm_top_p: Optional[float] = None
    llm_top_k: Optional[int] = None
    llm_frequency_penalty: Optional[float] = None
    llm_presence_penalty: Optional[float] = None

    agent_system_prompt: Optional[str] = None
    allow_web_search: Optional[bool] = None
    allow_vision: Optional[bool] = None  # ADICIONADO
    vision_model: Optional[str] = None   # ADICIONADO
    is_hyde_enabled: Optional[bool] = None

    tools_config: Optional[Dict[str, Any]] = None
    widget_config: Optional[Dict[str, Any]] = None
    reasoning_effort: Optional[str] = None
    verbosity: Optional[str] = None
    security_settings: Optional[Dict[str, Any]] = None

    is_subagent: Optional[bool] = None
    allow_direct_chat: Optional[bool] = None

class AgentResponse(AgentBase):
    id: UUID
    company_id: UUID
    created_at: datetime
    updated_at: datetime

    # Campos calculados (apenas WhatsApp mantido)
    has_whatsapp: bool = False

    class Config:
        from_attributes = True
